cffdump/cffdec.c

```diff
@@ -1256,35 +1287,55 @@ cp_load_state(uint32_t *dwords, uint32_t sizedwords, int level)
 {
 	enum shader_t stage;
 	enum state_t state;
+	enum state_src_t src;
 	uint32_t num_unit = (dwords[0] >> 22) & 0x1ff;
 	uint64_t ext_src_addr;
-	void *contents = NULL;
+	void *contents;
 	int i;
 
 	if (quiet(2) && !options->script)
 		return;
 
 	if (options->gpu_id >= 600)
-		a6xx_get_state_type(dwords, &stage, &state);
+		a6xx_get_state_type(dwords, &stage, &state, &src);
 	else if (options->gpu_id >= 400)
-		a4xx_get_state_type(dwords, &stage, &state);
+		a4xx_get_state_type(dwords, &stage, &state, &src);
 	else
-		a3xx_get_state_type(dwords, &stage, &state);
+		a3xx_get_state_type(dwords, &stage, &state, &src);
 
-	if (is_64b()) {
-		ext_src_addr = dwords[1] & 0xfffffffc;
-		ext_src_addr |= ((uint64_t)dwords[2]) << 32;
-		contents = dwords + 3;
-	} else {
-		ext_src_addr = dwords[1] & 0xfffffffc;
-		contents = dwords + 2;
+	switch (src) {
+	case STATE_SRC_DIRECT: ext_src_addr = 0; break;
+	case STATE_SRC_INDIRECT:
+		if (is_64b()) {
+			ext_src_addr = dwords[1] & 0xfffffffc;
+			ext_src_addr |= ((uint64_t)dwords[2]) << 32;
+		} else {
+			ext_src_addr = dwords[1] & 0xfffffffc;
+		}
+
+		break;
+	case STATE_SRC_BINDLESS: {
+		const unsigned base_reg =
+			stage == SHADER_COMPUTE ? regbase("HLSQ_CS_BINDLESS_BASE[0]") : regbase("HLSQ_BINDLESS_BASE[0]");
```

Comment:
hmm, is the bindless base always written before the `CP_LOAD_STATE` packet?  And is that a thing we can count on?  If not, I can't think of a very good option other than to defer dumping the state until draw time..

Reply:
Yes, because the bindless base is presumably what the hardware uses to actually prefetch the descriptor. If the bindless base weren't set beforehand then this packet wouldn't be useful at all. We're just doing the same computation here that HLSQ must be doing internally.

Reply:
Also, in practice the answer is yes, both with the Vulkan blob and with turnip.

Reply:
ok.. might be an interesting experiment to see what happens if we set the bindless base after the `CP_LOAD_STATE`.. that would at least confirm that prefetch happens immediately when CP sees this packet.

Anyways, looks reasonable, r-b